arbor/arbexcept.cpp

```diff
@@ -37,6 +37,11 @@ bad_global_property::bad_global_property(cell_kind kind):
     kind(kind)
 {}
 
+zero_thread_requested_error::zero_thread_requested_error(unsigned nbt):
+    arbor_exception(pprintf("threads must be a positive integer")),
```

Comment:
Since this PR seems to be restricted to pyarb changes, I think this exception would work better as a `pyarb_error` (defined in python/error.hpp)

Reply:
I see your point, but I'm not entirely happy with 1. this split anyway and 2. the particular contents of `python/error.hpp`. First of all, `python/pyarb.cpp` already contained one error translator `file_not_found_error`, which is a much clearer pattern: you register errors raised in C++ code, and translate them as you wish. Since the code of this particular change is C++ (and really all our code), even though restricted to being raised within Pyarb for the moment, it makes more sense to me to follow the same pattern. The `pyarb_error` in `python/error.hpp` should really be removed, it relies on implicit translation by `pybind11` and it's uses broken up into more meaningful exception classes.

If you agree, I can do that in a follow up PR. What do you think?

Reply:
The use of translation was a test and especially aimed at FileNotFound, because users expect FNF instead of RuntimeError. But yes, now that this is in for a while and caused no issues, we should port all our errors.

Reply:
I see what you're saying, but I don't think the solution is to move all of our custom errors to the arbor library. The errors that are used only in the `pyarb` library should be defined there, either in `python/error.hpp` or somewhere else. 
I agree that we should translate our exceptions into python exceptions whenever possible, but that has nothing to do with where the exceptions are defined. 
I think it's better to follow the current convention of using `pyarb_error` instead of `arbor_exception` in this case.

Reply:
There are already custom Arbor errors in `arbor/arbexcept.*pp` , right? They are not specific to a part of the API, else they would not live in one file. I don't think Python needs to be an exception (ha), other parts of Arbor may very well raise a `zero_thread_requested_error` at some point. I thought that having all errors in one place is so you can quickly find out if the exception you are raising already fits an existing class. 

Now, if the exception was very specific to Python I would agree with you, but this is about user input, which is not specific to Python.

I prefer to start (well, continue!) writing new exceptions in this pattern, and tackle existing `pyarb_error`s in the near future.

Reply:
We're looking at 2 libraries
1. `arbor` is the core arbor C++ library and has it's own exceptions in `arbor/arbexcept.*pp`.
2. `pyarb` is the python wrapper library, it needs `arbor` as a dependency but they are separate libraries. It has it's own exceptions in `python/error.*pp`.

An error that is only used in `pyarb` doesn't need to live in `arbor`. 
Separating the errors keeps the code clear, you can look at the type of the error if it's `pyarb_error` you know it's been raised from `pyarb`, if it's `arbor_exception` it's been raised from `arbor`. 
There's no need to accumulate all errors in the same file in the `arbor` lib. We can replicate what we have in the `arbor` lib in the `pyarb` lib as long as we keep that separation between the libraries clear.

Reply:
It is not a Python error, but an input error. It just happens to be only raised by the Python binding for the moment. In general, I think they should overlap as much as possible, not as little as possible.

I think "Python errors" should be seen as exceptions raised by Python machinery, e.g. C++ is unlikely to ever need a `TabError`.

Reply:
I think this is a small issue and I don't want to keep going around in circles. The current setup is to use `pyarb_error` in the `pyarb` library and I think this error belong there. It's up to you to change or not.